File: ruby/TensionsBetweenDifferentOOPrinciples/TensionBetweenSRPandLowCoupling/After/mini_robot.rb
require_relative 'plateau'

module TensionBetweenSRPandLowCoupling
  module After
    class MiniRobot
      # MiniRobot class now represents only a robot that accept commands to move in four directions inside the plateau.
      # So MiniRobot class follows the Single Responsibility Principle (SRP) while it has an additional dependency to 
      # the Plateau class. Because of it, coupling in higher.

      def initialize
        @plateau = Plateau.new
        @current_position = [5, 5]
      end

      def move_one_north
        north = [1, 0]
        @current_position = new_position north
      end

      def move_one_south
        south = [-1, 0]
        @current_position = new_position south
      end

      def move_one_east
        east = [0, 1]
        @current_position = new_position east
      end

      def move_one_west
        west = [0, -1]
        @current_position = new_position west
      end

      private def new_position(move)
        new_pos = [@current_position.first + move.first,
                   @current_position.last + move.last]

        if !@plateau.contains(new_pos)
          raise InvalidOperationException.new("Tentative to move outside plateau boundaries");
        end

        new_pos
      end
    end
  end
end

class InvalidOperationException < StandardError ; end

File: javascript/TensionsBetweenDifferentOOPrinciples/TensionBetweenSRPandLowCoupling/After/mini_robot.js
var Plateau = require("./plateau.js");

// MiniRobot class now represents only a robot that accept commands to move in four directions inside the plateau.
// So MiniRobot class follows the Single Responsibility Principle (SRP) while it has an additional dependency to
// the Plateau class. Because of it, coupling in higher.
var MiniRobot = function () {
  var plateau = new Plateau();

  this.currentPosition = [5, 5];

  this.newPosition = function (move) {
    var newPosition = [this.currentPosition[0] + move[0], this.currentPosition[1] + move[1]];

    if (!plateau.contains(newPosition))
      throw new Error("Tentative to move outside plateau boundaries");

    return newPosition;
  };
};

MiniRobot.prototype.moveOneNorth = function () {
  var north = [1, 0];
  this.currentPosition = this.newPosition(north);
};

MiniRobot.prototype.moveOneSouth = function () {
  var south = [-1, 0];
  this.currentPosition = this.newPosition(south);
};

MiniRobot.prototype.moveOneEast = function () {
  var east = [0, 1];
  this.currentPosition = this.newPosition(east);
};

MiniRobot.prototype.moveOneWest = function () {
  var west = [0, -1];
  this.currentPosition = this.newPosition(west);
};

module.exports = MiniRobot;
